examples/nats.md

```diff
@@ -0,0 +1,62 @@
+# Using the NATS backend
+
+[NATS](https://nats.io) is a high-performance connective technology supporting a spectrum of messaging, streaming, and materialization capabilities.
+
+One supported materialization a key-value store which is what is used in the implementation of the KINE backend.
+
+There are two options for integrating NATS:
+
+- As an external server or cluster KINE connects to
+- As an embedded server within KINE itself
+
+## Configuring KINE
+
+This is done by specifying the `--endpoint` option having the following format where everything but the `nats://` scheme is optional.
+
+```
+nats://[[[<auth>@]<host>]:<port>][?<params>]`
+```
+
+The above tokens are defined as follows:
+
+- `auth` - Inline username and password or standlaone token. Defaults is nothing.
+- `host` - Hostname or IP address the NATS server is addressed to. Default is `127.0.0.1`.
+- `port` - Port to bind the connection with the NATS server. Default is `4222`.
+- `params` - Query parameters which are listed below with defaults.
+
+The following query parameters are supported:
+
+- `bucket` - Specifies the name of the NATS key-value bucket. Default is `kine`.
+- `replicas` - Specifies the number of replicas for the bucket. Default is `1`.
+- `revHistory` - Specifies the number of revisions to keep in history. Default is `10`.
+- `slowMethod` - Specifies the duration of a method before it is considered slow. Default is `500ms`.
+- `contextFile` - Specifies the path to a NATS context file. If this is provided, the `auth` and `host` should not be provided. See the available [options](https://docs.nats.io/using-nats/nats-tools/nats_cli#configuration-contexts). Default is nothing.
+- `embedServer` - Specifies whether to embed a NATS server rather than relying on an external one. Default is `false`.
+- `serverConfig` - Specifies the path to a NATS server configuration file if `embedServer` is `true`. Defaults to nothing.
+- `dontListen` - Specifies whether the embedded NATS server should not rely on the TCP stack for a local client connection. Defaults to `false`
+- `stdoutLogging` - Specifies whether to log to STDOUT if `embedServer` is `true`. Default is false.
+
+## Examples
+
+### External
+
+- `nats://`
+- `nats://localhost:4222?replicas=3`
+- `nats://?contextFile=/path/to/context.json`
+- `nats://user:pass@localhost:4222`
+- `nats://token@localhost`
+
+Multiple URLs can be passed in a comma separated format, however only the first URL
+in the list will be evaluated for query parameters. For example:
+
+```
+nats://nats-1.example.com?bucket=k3s,nats://nats-2.example.com,nats://nats-3.example.com
+```
+
+However, if a `contextFile` is feasible to specify, this is recommended rather than an inline comma-separated list.
+
+### Embedded
+
+- `nats://?embedServer`
+- `nats://?embedServer&serverConfig=/path/to/server.conf`
+- `nats://?embedServer&dontListen`
```

Comment:
Thanks for adding docs!

Could you expand a bit on why someone might want to use the `dontListen` option? What's the default listen address and port? If it does listen, can other nodes connect to this node as the external NATS URL?

Reply:
I can add to the doc, but will explain here first for context. We have had NATS users who need to embed the server and their environment is constrained to only exposing a controlled or limited set of ports. So they take advantage of using this non-TCP approach to establishing a connection between a local client and server in the same process.

Reply:
sure that makes sense. I guess my question is, what is listening useful for? If you can't connect from another node when using the embedded server, why listen at all?

Reply:
Right, this would only be useful in a single node setup embedded k3s -> kine -> nats.. which may be very niche, but an option.

The more common setup is embedded that does listen and can cluster and/or be configured for specialized setups (the two-node use case I helped work on during the SUSE Hackathon for instance). The specialized setup will require additional opt-in custom code for NATS, but will enable for active-passive or active-active setups.

Reply:
Hey @bruth 

what is the embedded nats can be optionally put in leaf node and paired with a global nats cluster ? 

Then maybe devs can have their cake and eat it too:

either run in embedded standalone mode 

OR

run in embedded leaf mode paired up to a global nats cluster like Synadia etc .

is this a sensible option I wonder ?

Reply:
Yes you could configure it that way by passing a config file that declares the `leafnodes` block pointing to leaf node *host*. Other than the single KV bucket used in the backend, none of the other NATS capabilities are used. However, if you had a reason to mirror that KV back up to the hub cluster or even another leaf, that could definitely be done.

I have not tried it, but in theory you could have a _read-only_ clone of another k3s+kine instance that uses a mirror of another instance. The second instance would receive all the KV changes over time (since its a stream under the hood) so the other k3s instance would mirror the same state (albeit lagging a tad since the replication is asynchronous).

Reply:
Assuming I understand correctly, if you wanted three kine instances with embedded nats, they could be clustered together via nats. If you wanted a kine instance/cluster hosting a mirror from another instance/cluster that is fine to.

The cluster is only necessary for the HA/FT, but you could have a single instance origin and single instance mirror if desired.